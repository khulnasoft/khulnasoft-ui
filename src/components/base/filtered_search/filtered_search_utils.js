import first from 'lodash/first';
import last from 'lodash/last';
import isString from 'lodash/isString';
import { modulo } from '../../../utils/number_utils';

export const TERM_TOKEN_TYPE = 'filtered-search-term';

export const INTENT_ACTIVATE_PREVIOUS = 'intent-activate-previous';

export const TOKEN_CLOSE_SELECTOR = '.gl-token-close';

export function isEmptyTerm(token) {
  return token.type === TERM_TOKEN_TYPE && token.value.data.trim() === '';
}

export function normalizeTokens(tokens) {
  const result = [];
  tokens.forEach((token) => {
    if (isEmptyTerm(token)) {
      return;
    }

    if (token.type !== TERM_TOKEN_TYPE) {
      result.push({ ...token });
    } else if (result.length > 0 && typeof result[result.length - 1] === 'string') {
      result[result.length - 1] += ` ${token.value.data}`;
    } else {
      result.push(token.value.data);
    }
  });
  return result;
}

function assertValidTokens(tokens) {
  if (!Array.isArray(tokens) && !typeof tokens === 'string') {
    throw new TypeError('Either string or array of tokens is expected');
  }
}

export function needDenormalization(tokens) {
  if (typeof tokens === 'string') {
    return true;
  }

  assertValidTokens(tokens);

  return tokens.some((t) => typeof t === 'string' || !t.id);
}

/**
 * Given an initial index, step size and array length, returns an index that is
 * within the array bounds (unless step is 0; see † below).
 *
 * The step can be any positive or negative integer, including zero.
 *
 * An out-of-bounds index is considered 'uninitialised', and is handled
 * specially. For instance, the 'next' index of 'uninitialised' is the first
 * index:
 *
 *     stepIndexAndWrap(-1, 1, 5) === 0
 *
 * The 'previous' index of 'uninitialised' is the last index:
 *
 *     stepIndexAndWrap(-1, -1, 5) === 4
 *
 * †: If step is 0, the index is returned as-is, which may be out-of-bounds.
 *
 * @param {number} index The initial index.
 * @param {number} step The amount to step by (positive or negative).
 * @param {number} length The length of the array.
 * @returns {number}
 */
export function stepIndexAndWrap(index, step, length) {
  if (step === 0) return index;

  let start;
  const indexInRange = index >= 0 && index < length;

  if (indexInRange) {
    // Step from the valid index.
    start = index;
  } else if (step > 0) {
    // Step forwards from the beginning of the array.
    start = -1;
  } else {
    // Step backwards from the end of the array.
    start = length;
  }

  return modulo(start + step, length);
}

/**
 * Transforms a given token definition to an option definition.
 *
 * @param {Object} token A token definition (see GlFilteredSearch's
 *     availableTokens prop).
 * @returns {Object} A option definition (see GlFilteredSearchTokenSegment's
 *     options prop).
 */
export function tokenToOption({ icon, title, type, optionComponent }) {
  return { icon, title, value: type, component: optionComponent };
}

let tokenIdCounter = 0;
const getTokenId = () => {
  const tokenId = `token-${tokenIdCounter}`;
  tokenIdCounter += 1;
  return tokenId;
};
/**
 * Ensure the given token has an `id` property, which `GlFilteredSearch` relies
 * on as a unique key for the token.
 *
 * If the given token does not have an `id`, it returns a shallow copy of the
 * token with an `id`. Otherwise, it returns the given token.
 *
 * @param {object} token The token to check.
 * @returns {object} A token with an `id`.
 */
export function ensureTokenId(token) {
  if (!token.id) {
    return {
      ...token,
      id: getTokenId(),
    };
  }

  return token;
}

export function createTerm(data = '') {
  return {
    id: getTokenId(),
    type: TERM_TOKEN_TYPE,
    value: { data },
  };
}

export function denormalizeTokens(inputTokens, termsAsTokens = false) {
  assertValidTokens(inputTokens);

  const tokens = Array.isArray(inputTokens) ? inputTokens : [inputTokens];

  return tokens.reduce((result, t) => {
    if (typeof t === 'string') {
      if (termsAsTokens) {
        const trimmedText = t.trim();
        if (trimmedText) result.push(createTerm(trimmedText));
      } else {
        const stringTokens = t.split(' ').filter(Boolean);
        stringTokens.forEach((strToken) => result.push(createTerm(strToken)));
      }
    } else {
      result.push(ensureTokenId(t));
    }

    return result;
  }, []);
}

/**
 * Returns `true` if `text` contains `query` (case insensitive).
 *
 * This is used in `filter` and `find` array methods for token segment options.
 *
 * @param {string} text The string to look within.
 * @param {string} query The string to find inside the text.
 * @returns {boolean}
 */
export function match(text, query) {
  return text.toLowerCase().includes(query.toLowerCase());
}

export const termTokenDefinition = {
  type: TERM_TOKEN_TYPE,
  icon: 'title',
  title: 'Search for this text',
};

export function splitOnQuotes(str) {
  if (first(str) === "'" && last(str) === "'") {
    return [str];
  }

  if (first(str) === '"' && last(str) === '"') {
    return [str];
  }

  const queue = str.split(' ');
  const result = [];
  let waitingForMatchingQuote = false;
  let quoteContent = '';

  while (queue.length) {
    const part = queue.shift();
    const quoteIndex = part.indexOf('"');
    if (quoteIndex === -1) {
      if (waitingForMatchingQuote) {
        quoteContent += ` ${part}`;
      } else {
        result.push(part);
      }
    } else {
      const [firstPart, secondPart] = part.split('"', 2);

      if (waitingForMatchingQuote) {
        waitingForMatchingQuote = false;
        quoteContent += ` ${firstPart}"`;
        result.push(quoteContent);
        quoteContent = '';
        if (secondPart.length) {
          queue.unshift(secondPart);
        }
      } else {
        waitingForMatchingQuote = true;
        if (firstPart.length) {
          result.push(firstPart);
        }
        quoteContent = `"${secondPart}`;
      }
    }
  }
  return result;
}

/**
 *  wraps the incoming token in double quotes.
 *  Eg. Foo Bar becomes "Foo Bar"
 *
 *  1. token must have space.
 *  2. token should not already have a quote around it.
 */
export function wrapTokenInQuotes(token) {
  if (!isString(token)) {
    return token;
  }

  if (!token.includes(' ')) {
    return token;
  }

  const quotes = ["'", '"'];

  // If the token starts and ends with a quote, eg. "Foo Bar", then return the original token.
  if (quotes.some((quote) => first(token) === quote && last(token) === quote)) {
    return token;
  }

  return `"${token}"`;
}
