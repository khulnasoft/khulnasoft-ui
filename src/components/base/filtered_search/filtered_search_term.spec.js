import { nextTick } from 'vue';
import { shallowMount } from '@vue/test-utils';
import GlToken from '../token/token.vue';
import FilteredSearchTerm from './filtered_search_term.vue';
import { INTENT_ACTIVATE_PREVIOUS, TERM_TOKEN_TYPE } from './filtered_search_utils';

const availableTokens = [
  { type: 'foo', title: 'test1-foo', token: 'stub', icon: 'eye' },
  { type: 'bar', title: 'test2-bar', token: 'stub', icon: 'eye' },
  { type: 'baz', title: 'test1-baz', token: 'stub', icon: 'eye' },
];

const pointerClass = 'gl-cursor-pointer';

describe('Filtered search term', () => {
  let wrapper;

  const searchInputAttributes = { 'data-testid': 'foo-bar', 'data-prop': 'foo-property' };

  const defaultProps = {
    availableTokens: [],
    cursorPosition: 'end',
  };

  const segmentStub = {
    name: 'gl-filtered-search-token-segment-stub',
    template: '<div><slot v-if="!active" name="view"></slot><slot name="suggestions"></slot></div>',
    props: [
      'searchInputAttributes',
      'isLastToken',
      'currentValue',
      'viewOnly',
      'options',
      'active',
    ],
  };

  const createComponent = ({ termsAsTokens = false, ...props } = {}) => {
    wrapper = shallowMount(FilteredSearchTerm, {
      propsData: { ...defaultProps, ...props },
      provide: {
        termsAsTokens: () => termsAsTokens,
      },
      stubs: {
        'gl-filtered-search-token-segment': segmentStub,
      },
    });
  };

  const findSearchInput = () => wrapper.find('input');
  const findTokenSegmentComponent = () => wrapper.findComponent(segmentStub);
  const findToken = () => wrapper.findComponent(GlToken);

  it('renders value in inactive mode', () => {
    createComponent({ value: { data: 'test-value' } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders token in inactive mode with termsAsTokens', () => {
    createComponent({ value: { data: 'test value' }, termsAsTokens: true });
    expect(findToken().exists()).toBe(true);
    expect(findToken().props().viewOnly).toBe(false);
    expect(findToken().classes()).toContain(pointerClass);
    expect(findToken().text()).toBe('test value');
  });

  it('renders nothing with value in active mode (delegates to segment)', () => {
    createComponent({ value: { data: 'test-value' }, active: true });
    expect(wrapper.text()).not.toContain('test-value');
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders input with placeholder if placeholder prop is provided', () => {
    createComponent({ placeholder: 'placeholder-stub' });
    expect(findSearchInput().attributes('placeholder')).toBe('placeholder-stub');
  });

  it('filters suggestions by input', async () => {
    createComponent({ availableTokens, active: true, value: { data: 'test1' } });

    await nextTick();

    expect(findTokenSegmentComponent().props('options')).toHaveLength(2);
  });

  it('suggests text search when termsAsTokens=true', async () => {
    createComponent({
      availableTokens,
      active: true,
      value: { data: 'foo' },
      termsAsTokens: true,
    });

    await nextTick();

    expect(findTokenSegmentComponent().props('options')).toEqual([
      expect.objectContaining({
        value: 'foo',
        title: 'test1-foo',
      }),
      expect.objectContaining({
        value: TERM_TOKEN_TYPE,
        title: 'Search for this text',
      }),
    ]);
  });

  it.each`
    originalEvent   | originalPayload    | emittedEvent    | payload
    ${'activate'}   | ${undefined}       | ${'activate'}   | ${undefined}
    ${'deactivate'} | ${undefined}       | ${'deactivate'} | ${undefined}
    ${'split'}      | ${'foo'}           | ${'split'}      | ${'foo'}
    ${'submit'}     | ${undefined}       | ${'submit'}     | ${undefined}
    ${'previous'}   | ${undefined}       | ${'previous'}   | ${undefined}
    ${'next'}       | ${undefined}       | ${'next'}       | ${undefined}
    ${'complete'}   | ${'foo'}           | ${'replace'}    | ${{ type: 'foo' }}
    ${'complete'}   | ${TERM_TOKEN_TYPE} | ${'complete'}   | ${undefined}
    ${'backspace'}  | ${undefined}       | ${'destroy'}    | ${{ intent: INTENT_ACTIVATE_PREVIOUS }}
  `(
    'emits $emittedEvent when token segment emits $originalEvent',
    async ({ originalEvent, originalPayload, emittedEvent, payload }) => {
      createComponent({ active: true, value: { data: 'something' } });

      findTokenSegmentComponent().vm.$emit(originalEvent, originalPayload);

      await nextTick();

      expect(wrapper.emitted(emittedEvent)).toHaveLength(1);

      if (payload !== undefined) {
        expect(wrapper.emitted(emittedEvent)[0][0]).toEqual(payload);
      }
    }
  );

  it('passes `searchInputAttributes`, `isLastToken`, `currentValue` & `viewOnly` props to `GlFilteredSearchTokenSegment`', () => {
    const isLastToken = true;
    const viewOnly = true;
    const currentValue = [
      { type: 'filtered-search-term', value: { data: 'something' } },
      { type: 'filtered-search-term', value: { data: '' } },
    ];

    createComponent({
      value: { data: 'something' },
      searchInputAttributes,
      isLastToken,
      currentValue,
      viewOnly,
    });

    expect(findTokenSegmentComponent().props()).toEqual(
      expect.objectContaining({
        searchInputAttributes,
        isLastToken,
        currentValue,
        viewOnly,
      })
    );
  });

  it('by default sets `viewOnly` to false on `GlFilteredSearchTokenSegment`', () => {
    createComponent();

    expect(findTokenSegmentComponent().props('viewOnly')).toBe(false);
  });

  it('sets viewOnly prop and removes pointer class on token when termsAsTokens=true', () => {
    createComponent({
      value: { data: 'foo' },
      viewOnly: true,
      termsAsTokens: true,
    });

    expect(findToken().props().viewOnly).toBe(true);
    expect(findToken().classes()).not.toContain(pointerClass);
  });

  it('adds `searchInputAttributes` prop to search term input', () => {
    createComponent({
      placeholder: 'placeholder-stub',
      searchInputAttributes,
    });

    expect(findSearchInput().attributes('data-prop')).toBe(searchInputAttributes['data-prop']);
  });

  it('activates and deactivates when the input is focused/blurred', async () => {
    createComponent({ placeholder: 'placeholder-stub' });
    await nextTick();

    expect(wrapper.emitted()).toEqual({});

    await findSearchInput().trigger('focusin');

    expect(wrapper.emitted()).toEqual({ activate: [[]] });

    await findSearchInput().trigger('focusout');

    expect(wrapper.emitted()).toEqual({ activate: [[]], deactivate: [[]] });
  });

  describe.each([true, false])('when `viewOnly` is %s', (viewOnly) => {
    beforeEach(() => {
      createComponent({ viewOnly, searchInputAttributes, placeholder: 'placeholder-stub' });
    });

    it(`${viewOnly ? 'adds' : 'does not add'} \`gl-bg-gray-10\` class to search term input`, () => {
      expect(findSearchInput().classes('gl-bg-gray-10')).toBe(viewOnly);
    });
  });
});
