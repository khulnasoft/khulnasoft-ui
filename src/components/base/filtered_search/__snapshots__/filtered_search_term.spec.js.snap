// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filtered search term renders nothing with value in active mode (delegates to segment) 1`] = `
<div
  class="gl-filtered-search-term gl-h-auto"
  data-testid="filtered-search-term"
>
  <div
    class="gl-filtered-search-term-token"
    cursor-position="end"
    is-term=""
    value="test-value"
  />
</div>
`;

exports[`Filtered search term renders value in inactive mode 1`] = `
<div
  class="gl-filtered-search-term gl-h-auto"
  data-testid="filtered-search-term"
>
  <div
    class="gl-filtered-search-term-token"
    cursor-position="end"
    is-term=""
    value="test-value"
  >
    test-value
  </div>
</div>
`;
